packages/iTwinUI-react/src/core/Table/actionHandlers/selectHandler.ts

```diff
@@ -56,6 +56,109 @@ export const onSelectHandler = <T extends Record<string, unknown>>(
   onSelect?.(selectedData, newState);
 };
 
+/**
+ * Handles selection when clicked on a row while shift is selected.
+ */
+export const onShiftSelectHandler = <T extends Record<string, unknown>>(
+  state: TableState<T>,
+  action: ActionType,
+  instance: TableInstance<T>,
+) => {
+  const resetSelection = () => {
+    instance.toggleAllRowsSelected(false);
+  };
+
+  const getGreaterId = (id: string) => {
+    const greaterId = id.split('.');
+    greaterId[greaterId.length - 1] = `${
+      parseInt(greaterId[greaterId.length - 1]) + 1
+    }`;
+    return greaterId.join('.');
+  };
+
+  const getParentId = (id: string) => {
+    const lastDotIndex = id.lastIndexOf('.');
+    if (lastDotIndex < 0) {
+      return null;
+    }
+    return id.substring(0, lastDotIndex);
+  };
+
+  /**
+   * @returns `true` if `subId` is indeed a sub-row of `id`, else `false`
+   * @example
+   * isSubRow('1.4', '1.4.2.3') // true
+   * isSubRow('1.3', '1.2.5')   // false
+   * isSubRow('2.4.5', '2.4.5') // true
+   */
+  const isSubRow = (id: string, subId: string) => {
+    const idSplit = id.split('.');
+    const subIdSplit = subId.split('.');
+
+    let i = 0;
+    while (i < idSplit.length) {
+      if (idSplit[i] !== subIdSplit[i]) {
+        return false;
+      }
+      i++;
+    }
+
+    return true;
+  };
+
+  /** Returns the next row as if it was just a non sub-rows list */
+  const getNextRow = (currentRow: Row<T>, endId: string) => {
+    let nextRow;
+    if (
+      (currentRow.subRows ?? []).length > 0 &&
+      isSubRow(currentRow.id, endId)
+    ) {
+      // child
+      nextRow = currentRow.subRows[0];
+    } else {
+      // next sibling
+      nextRow = instance.rowsById[getGreaterId(currentRow.id)];
+
+      // If the next sibling doesn't exist, go to the next child at a one upper depth
+      let parentId = getParentId(currentRow.id);
+      while (nextRow == null && parentId != null) {
+        nextRow = instance.rowsById[getGreaterId(parentId)];
+        parentId = getParentId(parentId);
+      }
+    }
+
+    return nextRow;
+  };
+
+  resetSelection();
+
+  // If no row selected before shift clicking, then assume first row is the last selectedRow (like Windows File Explorer)
+  let startId = state.lastSelectedRow ?? instance.rows[0].id;
```

Comment:
i think this logic is buggy because `state.lastSelectedRow` is not getting cleared after all rows are deselected.

to see what I mean, try selecting any random row and deselect it. Now try shift clicking any other row. It will behave differently from how it does when the page loads.

Reply:
I think that behavior is intended. I did it that way since Windows Explorer and other lists like Linux Files Explorer also did it the same way

[vokoscreenNG-2022-09-02_06-31-26.webm](https://user-images.githubusercontent.com/45748283/188121254-47ea828e-9b69-439d-a635-f268227e9bc2.webm)

Reply:
Ohh, yes, we could implement some parts our way too. But maybe we could also incorporate some parts from existing essential tools to give a more familiar and intuitive feel. Some reasons I feel the Windows Explorer does it that way:

1. If someone does a shift click after all rows are deselected without `lastSelectedRow` saved, it would be strange to see the same behavior as a normal click (because a normal click toggles the row, and if no `lastSelectedRow` is there, then shift click will also toggle the row). So instead of doing a toggle, we can do a real shift click from the `lastSelectedRow` to the current clicked row. 

2. If the user accidentally clicks outside the table (or similar) and loses their selection, they would have to reselect the first row. This might be a problem if the user went through a lot of effort to find the starting row the first time (e.g. finding the correct image when all images have similar names with random number timestamps 
	```
	  20220902-122553.jpg
	  20220902-123219.jpg
	  20220902-121637.jpg
	  ...
	  ```

	or if the file names are so long that the user has to make the column wider to find the correct name).

	  ```
	  abc<...............................>01.txt
	  abc<...............................>02.txt
	  abc<...............................>03.txt
	  ...
	  ```

	So if the `lastSelectedRow` is saved, the user can assume the start row is still selected and just move to finding the ending row. 

But I do see how saving the `lastSelectedRow` might be confusing to some users too.